File: src/day09/Day09.kt
package day09

import java.io.File

const val FILENAME = "src/day09/Input.txt"

val XMAS = File(FILENAME).readLines().map { it.toLong() }.toTypedArray()

fun part01(): Long {
    val preambleSize = 25
    val preamble = XMAS.sliceArray(0..preambleSize - 1).toMutableList()
    val remaining = XMAS.sliceArray(preambleSize..XMAS.size - 1).toMutableList()

    while (remaining.any()) {
        val n = remaining[0]
        var exists = false
        for (i in preamble) {
            if (preamble.contains(n - i)) {
                exists = true
                preamble.add(n)
                remaining.removeAt(0)
                preamble.removeAt(0)
                break
            }
        }
        if (!exists) {
            println(n)
            return n
        }
    }
    return -1
}

fun part02(): Long {
    val invalid = part01()
    for (i in 2..XMAS.size) {
        for (j in 0..XMAS.size - i) {
            val contiguous = XMAS.sliceArray(j..(j + i) - 1)
            if (contiguous.sum() == invalid)
                return contiguous.minOrNull()!! + contiguous.maxOrNull()!!
        }
    }
    return -1
}

fun main() {
    println("Advent of Code 2020, Day 09")
    println(part02()) // Calls part01() as well
}

File: day09/day09.go
package day09

import (
	"advent-of-code-2020/utils"
	"slices"
)

const filename = "./inputs/day09.txt"

var xmas = utils.MapToInt(utils.ReadLines(filename))

var invalid = make(chan int, 1)

func part01() int {
	preambleSize := 25
	preamble := xmas[:preambleSize]
	remaining := xmas[preambleSize:]

	for len(remaining) > 0 {
		n := remaining[0]
		exists := false
		for _, i := range preamble {
			if slices.Contains(preamble, n-i) {
				exists = true
				preamble = append(preamble, n)
				remaining = remaining[1:]
				preamble = preamble[1:]
				break
			}
		}
		if !exists {
			invalid <- n
			return n
		}
	}
	invalid <- -1
	return -1
}

func part02() int {
	invalid := <-invalid
	for i := 2; i <= len(xmas); i++ {
		for j := 0; j <= len(xmas)-i; j++ {
			contiguous := xmas[j:(j + i)]
			if utils.Sum(contiguous) == invalid {
				return slices.Min(contiguous) + slices.Max(contiguous)
			}
		}
	}
	return -1
}

func Main() (int, func() int, func() int) {
	return 9, part01, part02
}
